dali/pipeline/operators/util/external_source.cc

```diff
@@ -17,29 +17,46 @@
 namespace dali {
 
 template<>
-void ExternalSource<CPUBackend>::RunImpl(SampleWorkspace *ws) {
-  // Wrap the output tensor around our data
-  auto &output = ws->Output<CPUBackend>(0);
-  cudaStream_t stream = ws->has_stream() ? ws->stream() : 0;
-  if (data_in_tl_) {
-    DALI_ENFORCE(OperatorBase::batch_size_ == static_cast<int>(tl_data_.ntensor()),
-      "Data list provided to ExternalSource needs to have batch_size length.");
-    output.Copy(tl_data_, ws->data_idx(), stream);
-  } else {
-    DALI_ENFORCE(OperatorBase::batch_size_ == static_cast<int>(t_data_.size()),
-      "Data list provided to ExternalSource needs to have batch_size length.");
-    auto &data = t_data_[ws->data_idx()];
-    output.Copy(data, stream);
-  }
-
-  std::lock_guard<std::mutex> l(samples_processed_m_);
-  if (++samples_processed_ >= batch_size_) {
-    samples_processed_ = 0;
-    {
-      std::lock_guard<std::mutex> busy_lock(busy_m_);
-      busy_ = false;
+void ExternalSource<CPUBackend>::RunImpl(HostWorkspace *ws) {
+  bool is_tl_data;
+  uptr_tl_list_type tl_data;
```

Comment:
I find the names `tl_data` and `tl_data_` used for list of one element and a list of all cached elements respectively a bit confusing.

Reply:
Done